### PR TITLE
Introduce lifeVersion to IndexValue.

### PR DESCRIPTION
**Pull request by dharju**
PR to introduce lifeVersion to support undeletes, taken from https://github.com/linkedin/ambry/pull/1245 with comment fixes

### CODE REVIEW

File: ambry-store/src/main/java/com.github.ambry.store/IndexValue.java [hunk 1]
```diff
@@ -318,28 +374,30 @@ ByteBuffer getBytes() {
...
   }
 
   @Override
   public String toString() {
     return "Offset: " + offset + ", Size: " + getSize() + ", Deleted: " + isFlagSet(Flags.Delete_Index)
-        + ", TTL Updated: " + isFlagSet(Flags.Ttl_Update_Index) + ", ExpiresAtMs: " + getExpiresAtMs()
-        + ", Original Message Offset: " + getOriginalMessageOffset() + (version != PersistentIndex.VERSION_0 ? (
-        ", OperationTimeAtSecs " + getOperationTimeInMs() + ", AccountId " + getAccountId() + ", ContainerId "
-            + getContainerId()) : "");
+        + ", TTL Updated: " + isFlagSet(Flags.Ttl_Update_Index) + ", Undelete Updated: " + isFlagSet(
```
Comment (zzmao) [inline ambry-store/src/main/java/com.github.ambry.store/IndexValue.java]: Undelete Updated -> Undelete

**Comment by zzmao** [comment]
oops, there is a conflict.

File: ambry-store/src/main/java/com.github.ambry.store/IndexValue.java (hunk 1 above)
Comment (dharju) [reply ambry-store/src/main/java/com.github.ambry.store/IndexValue.java]: done

File: ambry-store/src/main/java/com.github.ambry.store/IndexValue.java [hunk 2]
```diff
@@ -318,28 +374,30 @@ ByteBuffer getBytes() {
...
+        Flags.Undelete_Index) + ", ExpiresAtMs: " + getExpiresAtMs() + ", Original Message Offset: "
+        + getOriginalMessageOffset() + (formatVersion != PersistentIndex.VERSION_0 ? (", OperationTimeAtSecs "
+        + getOperationTimeInMs() + ", AccountId " + getAccountId() + ", ContainerId " + getContainerId())
+        : "") + (formatVersion > PersistentIndex.VERSION_2 ? ", Update Version:" + lifeVersion : "");
```
[the remark this answers is not in the document]
Comment (dharju) [reply ambry-store/src/main/java/com.github.ambry.store/IndexValue.java]: done

File: ambry-store/src/main/java/com.github.ambry.store/IndexValue.java [hunk 3]
```diff
@@ -118,9 +130,29 @@
             : Utils.Infinite_Time;
         accountId = value.getShort();
         containerId = value.getShort();
+        lifeVersion = 0;
+        break;
+      case PersistentIndex.VERSION_3:
+        if (value.capacity() != INDEX_VALUE_SIZE_IN_BYTES_V3) {
+          throw new IllegalArgumentException("Invalid buffer size for formatVersion 1");
```
[the remark this answers is not in the document]
Comment (dharju) [reply ambry-store/src/main/java/com.github.ambry.store/IndexValue.java]: done

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 4]
```diff
@@ -1028,7 +1028,7 @@ private boolean copyRecords(LogSegment logSegmentToCopy, List<IndexEntry> srcInd
               } else {
                 IndexValue tgtValue = new IndexValue(srcValue.getSize(), fileSpan.getStartOffset(), srcValue.getFlags(),
                     srcValue.getExpiresAtMs(), srcValue.getOperationTimeInMs(), srcValue.getAccountId(),
-                    srcValue.getContainerId());
+                    srcValue.getContainerId(), (short) 0);
```
[the remark this answers is not in the document]
Comment (dharju) [reply ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: done

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 5]
```diff
@@ -1039,7 +1039,7 @@ private boolean copyRecords(LogSegment logSegmentToCopy, List<IndexEntry> srcInd
...
```
[the remark this answers is not in the document]
Comment (dharju) [reply ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: done

**Comment by dharju** [comment]
> Generally looks good to me.
> 
> My only concern is that `PersistentIndex.CURRENT_VERSION`, looks like a good time to make it configurable instead of hard coding.

I tried doing that, but the change is so deep I think I would need to do a separate PR for it

**Comment by zzmao** [comment]
> > Generally looks good to me.
> > My only concern is that `PersistentIndex.CURRENT_VERSION`, looks like a good time to make it configurable instead of hard coding.
> 
> I tried doing that, but the change is so deep I think I would need to do a separate PR for it

OK. Let's add it in next PR. Current PR looks good to me. Some travis error. Please fix them.

**Pull request merged**